contracts/Everlong.sol

```diff
@@ -73,72 +76,120 @@ contract Everlong is Admin, ERC4626, Positions {
     constructor(
         string memory name_,
         string memory symbol_,
         address hyperdrive_,
-        address asset_
+        bool asBase_,
+        uint256 targetIdleLiquidity_
     ) Admin() {
         // Store constructor parameters.
         _name = name_;
         _symbol = symbol_;
-        hyperdrive = hyperdrive_;
-        _asset = asset_;
+        _hyperdrive = hyperdrive_;
+        _asBase = asBase_;
+        _asset = _asBase
+            ? IHyperdrive(_hyperdrive).baseToken()
+            : IHyperdrive(_hyperdrive).vaultSharesToken();
+        _targetIdleLiquidity = targetIdleLiquidity_;
 
         // Attempt to retrieve the decimals from the {_asset} contract.
         // If it does not implement `decimals() (uint256)`, use the default.
-        (bool success, uint8 result) = _tryGetAssetDecimals(_asset);
+        (bool success, uint8 result) = _tryGetAssetDecimals(asset());
         _decimals = success ? result : _DEFAULT_UNDERLYING_DECIMALS;
+
+        // Give max approval for `_asset` to the hyperdrive contract.
+        IERC20(_asset).approve(hyperdrive_, type(uint256).max);
     }
 
-    /// @dev Address of the underlying Hyperdrive instance.
+    function hyperdrive() public view returns (address) {
+        return _hyperdrive;
+    }
+
+    /// @notice Address of the underlying Hyperdrive instance.
     /// @dev MUST be an ERC20 token contract.
-    /// @dev MUST NOT revert.
+    /// @return Hyperdrive address.
     function asset() public view virtual override returns (address) {
         return _asset;
     }
 
-    /// @dev Returns the name of the Everlong token.
+    /// @notice Returns the name of the Everlong token.
+    /// @return Everlong token name.
     function name() public view virtual override returns (string memory) {
         return _name;
     }
 
-    /// @dev Returns the symbol of the Everlong token.
+    /// @notice Returns the symbol of the Everlong token.
+    /// @return Everlong token symbol.
     function symbol() public view virtual override returns (string memory) {
         return _symbol;
     }
 
-    /// @dev Returns the kind of the Everlong instance.
+    /// @notice Returns the kind of the Everlong instance.
+    /// @return Everlong contract kind.
     function kind() public view virtual returns (string memory) {
         return "Everlong";
     }
 
-    /// @dev Returns the version of the Everlong instance.
+    /// @notice Returns the version of the Everlong instance.
+    /// @return Everlong contract version.
     function version() public view virtual returns (string memory) {
         return "v0.0.1";
     }
 
+    /// @inheritdoc IRebalancing
+    function canRebalance() public view returns (bool) {
+        return
+            hasMaturedPositions() ||
+            IERC20(_asset).balanceOf(address(this)) >=
+            IHyperdrive(_hyperdrive).getPoolConfig().minimumTransactionAmount;
```

Comment:
Maybe in the future, for gas cost reasons. This can be a configuration parameter set in the vault itself and can default to the minimumTransactionAmount from hyperdrive.

Reply:
Yeah v true, this code is def not optimized. Will leave a todo